net: packet clone - add copying of missing flags inside clone_pkt_attributes() function.

MarGasiorek opened the pull request:
Add cloning of missing flags (attributes).
Update UT.
Small code improvement.

MarGasiorek committed:
net: Follow coding style when converting int to bool

Added double exclamation mark where missing in order to keep
code in one convention.

Signed-off-by: Marcin Gasiorek <marcin.gasiorek@nordicsemi.no>

MarGasiorek committed:
tests: net: Add missing tests to net_pkt_clone

Add tests for flag cloning and couple extra.

Signed-off-by: Marcin Gasiorek <marcin.gasiorek@nordicsemi.no>

jukkar commented on include/zephyr/net/net_pkt.h:
```diff
@@ -509,7 +509,7 @@ static inline void net_pkt_set_eof(struct net_pkt *pkt, bool eof)
 #if defined(CONFIG_NET_ROUTE)
```
The commit subject `net: Code improvement` is too generic. We could change that to
`net: Follow coding style when converting int to bool`

MarGasiorek replied:
Sorry for that, I changed comment as you suggest.

jukkar commented on tests/net/net_pkt/prj.conf:
```diff
@@ -4,11 +4,13 @@ CONFIG_ZTEST_NEW_API=y
 CONFIG_ENTROPY_GENERATOR=y
 CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_NET_IPV6_NBR_CACHE=y
+CONFIG_NET_ETHERNET_BRIDGE=y
```
Why do we need bridging enabled?

MarGasiorek replied:
I enabled it in order to test l2_bridged flag:
https://github.com/zephyrproject-rtos/zephyr/blob/d42d14e392e70a846abe5baaceb849ef21a50584/include/zephyr/net/net_pkt.h#L394

jukkar replied:
Ok, was just wondering about that because it was not very clear.

rlubos commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1809,6 +1809,13 @@ static void clone_pkt_attributes(struct net_pkt *pkt, struct net_pkt *clone_pkt)
 	net_pkt_set_priority(clone_pkt, net_pkt_priority(pkt));
 	net_pkt_set_orig_iface(clone_pkt, net_pkt_orig_iface(pkt));
 	net_pkt_set_captured(clone_pkt, net_pkt_is_captured(pkt));
+	net_pkt_set_sent(clone_pkt, net_pkt_sent(pkt));
+	net_pkt_set_queued(clone_pkt, net_pkt_queued(pkt));
```
Is this correct to clone those flags? If the original packet was marked as sent/queued, it does not seem to be true for the clone.

Anyway, both these flags seem no longer in use (they aren't read anywhere within the codebase). The setters are called in a couple of places, but according to the comments/git messages they seemed to have been used by the old TCP implementation (some comments even refer to obsolete TCP code).

Maybe if we're doing cleanup in the flags, we should consider removing them instead?

jukkar replied:
> Is this correct to clone those flags? If the original packet was marked as sent/queued, it does not seem to be true for the clone.

I was thinking that too, but it is difficult to know what flags would not need to be cloned. The cloning implies also that we get identical copy of the original net_pkt so with in mind I considered it ok if we just copy all the flags.

> 
> Anyway, both these flags seem no longer in use (they aren't read anywhere within the codebase). The setters are called in a couple of places, but according to the comments/git messages they seemed to have been used by the old TCP implementation (some comments even refer to obsolete TCP code).

The sent flag is shared with eof, so while we could remove the function, the actual bit field cannot be removed (but could be renamed).

> 
> Maybe if we're doing cleanup in the flags, we should consider removing them instead?

Yes, that is good idea.

MarGasiorek replied:
> 

> Maybe if we're doing cleanup in the flags, we should consider removing them instead?

> Yes, that is good idea.

If this flag is really unused then maybe should be renamed to `unused` for future use.

MarGasiorek replied:
I removed all deprecated and unused flags together with setters/getters.

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1809,6 +1809,13 @@ static void clone_pkt_attributes(struct net_pkt *pkt, struct net_pkt *clone_pkt)
+	net_pkt_set_queued(clone_pkt, net_pkt_queued(pkt));
+	net_pkt_set_ptp(clone_pkt, net_pkt_is_ptp(pkt));
+
+#if defined(CONFIG_NET_ROUTE)
```
Good point, +1 to that.

MarGasiorek replied:
The `#if defined(CONFIG_NET_ROUTE)` has been removed.

jukkar reviewed (commented):
I would prefer that the commit "net: ip: Add cloning of missing attributes" is split to two. First one that does the removal of things, and the second one that does the cloning.

MarGasiorek committed:
net: Deprecated and unused flags cleanup.

In net_pkt structure couple of flags are no longer used.
Flag pkt_queued can be removed permanently togeter with setter
and getter functions.
Second flag sent_or_eof has been renamed because it is still
used partially only for indicating EOF. Additionally unused setter
and getter for 'sent_' part of this flag are removed.

Signed-off-by: Marcin Gasiorek <marcin.gasiorek@nordicsemi.no>

MarGasiorek committed:
net: Remove '#ifdef CONFIG_NET_ROUTE' around forwarding getter and setter.

Because `forwarding` flag is available unconditionally the compilation
guard has no make sense. Additionally this flag and related getter and
setter can be used for e.g. IPv4 packets forwarding in future, which
would be impossible looking at `NET_ROUTE` dependencies.

Signed-off-by: Marcin Gasiorek <marcin.gasiorek@nordicsemi.no>

MarGasiorek committed:
net: ip: Add cloning of missing attributes

A few attributes isn't cloned togerder with pkt.
This commit add missing part.
Additionally, because the `eof` flag in net_pkt structure is
avilable unconditionally hence `#if defined(CONFIG_NET_SOCKETS)`
guard has been removed form setter and getter functions for
this flag.

Signed-off-by: Marcin Gasiorek <marcin.gasiorek@nordicsemi.no>

MarGasiorek commented:
> I would prefer that the commit "net: ip: Add cloning of missing attributes" is split to two. First one that does the removal of things, and the second one that does the cloning.

Done

rlubos reviewed (approved):
Thanks
